### PR TITLE
Rework backtrace reader into an iterator

### PR DESCRIPTION
**Pull request by bugadani**
This impl, if I didn't break RISC-V should be much more flexible than before: we can print more frames, and we don't immediately stop if we encounter a recursive function. Also, we can remove some duplicate code.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -99,21 +85,79 @@ fn exception_handler(context: &arch::TrapFrame) -> ! {
...
-            println!("0x{:x}", addr);
+}
+
+impl Iterator for Backtrace {
+    type Item = usize;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        while self.depth < MAX_BACKTRACE_ADRESSES {
+            let address = self.read_next_pc();
+            if address == 0 {
```
Comment (bugadani) [inline src/lib.rs]: This is incorrect on Xtensa, we `| 0x4000_0000`

**Comment by bjoernQ** [comment]
Mhhh - this test on ESP32

With esp-backtrace from crates.io
```
!! A panic occured in 'examples\hello_world.rs', at line 60, column 5

PanicInfo {
    payload: Any { .. },
    message: Some(
        explicit panic,
    ),
    location: Location {
        file: "examples\\hello_world.rs",
        line: 60,
        col: 5,
    },
    can_unwind: true,
}

Backtrace:

0x400e80f7
0x400e80f7 - core::panicking::panic
    at C:\Users\Bjoern\.rustup\toolchains\esp\lib\rustlib\src\rust\library\core\src\panicking.rs:117
0x400d4621
0x400d4621 - hello_world::foo5
    at C:\projects\esp\esp-hal\esp32-hal\examples\hello_world.rs:60
0x400d460d
0x400d460d - hello_world::foo4
    at C:\projects\esp\esp-hal\esp32-hal\examples\hello_world.rs:57
0x400d4601
0x400d4601 - hello_world::foo3
    at C:\projects\esp\esp-hal\esp32-hal\examples\hello_world.rs:53
0x400d45f5
0x400d45f5 - hello_world::foo2
    at C:\projects\esp\esp-hal\esp32-hal\examples\hello_world.rs:49
0x400d45e9
0x400d45e9 - hello_world::foo1
    at C:\projects\esp\esp-hal\esp32-hal\examples\hello_world.rs:45
0x400d45dd
0x400d45dd - hello_world::foo
    at C:\projects\esp\esp-hal\esp32-hal\examples\hello_world.rs:41
0x400d46d2
0x400d46d2 - hello_world::__xtensa_lx_rt_main
    at C:\projects\esp\esp-hal\esp32-hal\examples\hello_world.rs:35
0x400e42c3
0x400e42c3 - Reset
    at C:\Users\Bjoern\.cargo\registry\src\index.crates.io-6f17d22bba15001f\xtensa-lx-rt-0.16.0\src\lib.rs:70
0x400dcef1
0x400dcef1 - ESP32Reset
    at C:\projects\esp\esp-hal\esp32-hal\src\lib.rs:88
```

With this
```
Hello world!


!! A panic occured in 'examples\hello_world.rs', at line 60, column 5

PanicInfo {
    payload: Any { .. },
    message: Some(
        explicit panic,
    ),
    location: Location {
        file: "examples\\hello_world.rs",
        line: 60,
        col: 5,
    },
    can_unwind: true,
}

Backtrace:





```

🤔

**Comment by bugadani** [comment]
Well I did expect some breakage, but for RISC-V. I never thought testing on the S3 isn't indicative of the old ESP32 🥲

File: src/lib.rs [hunk 2]
```diff
@@ -51,16 +45,8 @@ unsafe fn __user_exception(cause: arch::ExceptionCause, context: arch::Context)
...
-    for e in backtrace {
-        if let Some(addr) = e {
-            println!("0x{:x}", addr);
-        }
-    }
-
-    println!("");
-    println!("");
-    println!("");
+    let backtrace = Backtrace::new(core::ptr::addr_of!(context) as u32, 0);
```
Comment (bugadani) [inline src/lib.rs:48]: Otherwise exception handling skips the topmost frame

File: src/riscv.rs [hunk 3]
```diff
@@ -101,60 +101,31 @@ MTVAL=0x{:08x}
     }
 }
 
-/// Get an array of backtrace addresses.
+/// Get an iterator of backtrace addresses.
 ///
 /// This needs `force-frame-pointers` enabled.
-pub fn backtrace() -> [Option<usize>; MAX_BACKTRACE_ADRESSES] {
+#[inline(never)]
```
Comment (bugadani) [inline src/riscv.rs:107]: Optimizer doing its own thing could make things unpredictable, so by disallowing inlining I hope we always skip the correct number of frames

File: src/lib.rs [hunk 4]
```diff
@@ -99,21 +85,76 @@ fn exception_handler(context: &arch::TrapFrame) -> ! {
...
+}
+
+impl Backtrace {
+    #[inline(always)]
+    fn new(fp: u32, suppress: usize) -> Self {
+        let mut this = Backtrace { fp, depth: 0 };
+
+        // We need to skip the first few frames here, otherwise returning to the caller might cause
+        // problems.
+        for _ in 0..suppress {
```
Comment (bugadani) [inline src/lib.rs:110]: This is where the magic happens. Looks like returning from `backtrace` pops the first stack frame we want to read 😅

**Review by bugadani — COMMENTED** [review]
eww. and it still doesn't work

**Pull request closed without merging**